From 847bc4fa699b5b1c2517656c6755e44e952bbfc0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stephan=20M=C3=BCller?= <smueller@chronox.de>
Date: Sun, 21 Nov 2021 15:51:44 +0100
Subject: [PATCH 0076/1646] [Backport] crypto: dh - limit key size to 2048 in
 FIPS mode

mainline inclusion
from mainline-v5.17-rc1
commit 1e146c393b152a31771b49af5d104d9ed846da9b
category: feature
bugzilla: 188668

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=1e146c393b152a31771b49af5d104d9ed846da9b

--------------------------------

FIPS disallows DH with keys < 2048 bits. Thus, the kernel should
consider the enforcement of this limit.

Signed-off-by: Stephan Mueller <smueller@chronox.de>
Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
Signed-off-by: Cui GaoSheng <cuigaosheng1@huawei.com>
---
 crypto/dh.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/crypto/dh.c b/crypto/dh.c
index cd4f32092e5c..38557e64b4b3 100644
--- a/crypto/dh.c
+++ b/crypto/dh.c
@@ -5,6 +5,7 @@
  * Authors: Salvatore Benedetto <salvatore.benedetto@intel.com>
  */
 
+#include <linux/fips.h>
 #include <linux/module.h>
 #include <crypto/internal/kpp.h>
 #include <crypto/kpp.h>
@@ -47,6 +48,9 @@ static inline struct dh_ctx *dh_get_ctx(struct crypto_kpp *tfm)
 
 static int dh_check_params_length(unsigned int p_len)
 {
+	if (fips_enabled)
+		return (p_len < 2048) ? -EINVAL : 0;
+
 	return (p_len < 1536) ? -EINVAL : 0;
 }
 
-- 
2.17.1

